mitmproxy/addons/proxyserver.py

```diff
@@ -222,156 +151,96 @@ def configure(self, updated):
-                        set_instance(instance)
-                        # TODO: This is a bit confusing currently for `-p 0`.
-                        addrs = {
-                            f"{human.format_address(s.getsockname())}"
-                            for s in instance.sockets
-                        }
-                        ctx.log.info(
-                            f"{name} server listening at {' and '.join(addrs)}"
-                        )
+            if any(isinstance(m, mode_specs.TransparentMode) for m in modes):
+                if platform.original_addr:
+                    platform.init_transparent_mode()
+                else:
+                    raise exceptions.OptionsError("Transparent mode not supported on this platform.")
 
-    async def shutdown_server(self):
-        for name, instance, set_instance, _, _ in self._server_desc:
-            if instance is not None:
-                ctx.log.info(f"Stopping {name} server...")
-                try:
-                    instance.close()
-                    await instance.wait_closed()
-                except OSError as e:
-                    ctx.log.error(str(e))
+            if self.is_running:
+                asyncio.create_task(self.setup_servers())
+
+    async def setup_servers(self) -> bool:
+        all_ok = True
+        async with self._lock:
+            new_servers: dict[str, ServerInstance] = dict.fromkeys(ctx.options.mode)  # type: ignore
+            if not ctx.options.server:
+                new_servers.clear()
+
+            # Shutdown modes that have been removed from the list.
+            shutdown_tasks = [
+                s.stop() for spec, s in self.servers.items()
+                if spec not in new_servers
+            ]
+            for ret in await asyncio.gather(*shutdown_tasks, return_exceptions=True):
+                if ret:
+                    all_ok = False
+                    ctx.log.error(str(ret))
+
+            new_instances: list[ServerInstance] = []
+            for spec in new_servers:
+                if existing := self.servers.get(spec, None):
+                    new_servers[spec] = existing
                 else:
-                    set_instance(None)
+                    instance: ServerInstance = ServerInstance.make(spec, self)
+                    new_instances.append(instance)
+                    new_servers[spec] = instance
 
-    async def handle_connection(self, connection_id: tuple):
-        handler = self._connections[connection_id]
-        task = asyncio.current_task()
-        assert task
-        asyncio_utils.set_task_debug_info(
-            task,
-            name=f"Proxyserver.handle_connection",
-            client=handler.client.peername,
-        )
-        try:
-            await handler.handle_client()
-        finally:
-            del self._connections[connection_id]
+            for ret in await asyncio.gather(*[m.start() for m in new_instances], return_exceptions=True):
```

Comment:
Could there be the case where a server instance needs to be notified of a configuration change, where the mode stays the same but other settings get changed? If so maybe do something like
```python
            new_instances: list[ServerInstance] = []
            existing_instances: list[ServerInstance] = []
            for spec in new_servers:
                if existing := self.servers.get(spec, None):
                    existing_instances.append(instance)
                    new_servers[spec] = existing
                else:
                    instance: ServerInstance = ServerInstance.make(spec, self)
                    new_instances.append(instance)
                    new_servers[spec] = instance

            for ret in await asyncio.gather(
                *[m.start() for m in new_instances],
                *[m.reconfigure() for m in existing_instances],
                return_exceptions=True,
            ):
                if ret:
                    all_ok = False
                    ctx.log.error(str(ret))
```